django/models/README.md

```diff
@@ -0,0 +1,243 @@
+    first_name = models.CharField(max_length=100)
+    # No : Repeat model name person
+    person_first_name = models.CharField(max_length=100)
+    
+    # Yes: Many to Many in plural form
+    companies = model.ManyToMany(to='company.Company', blank=True, related_name='people')
+
+    # Yes
+    activated = models.BooleanField()
+    # No
+    is_active = models.BooleanFiled()
+```
+
+### Relationship
+
+- When defining foreign key, one to one or many to many in a model it should be
+lazy reference instead of importing the model class directly, preventing
+recursive import exception.
+[Read more](https://docs.djangoproject.com/en/3.1/ref/models/fields/#foreignkey)
+- Always add related name parameter to foreign key or many to many field.
+
+```python
+# Yes
+class Employee(models.Model):
+    person = models.ForeignKey(to='people.Person', on_delete=models.CASCADE, related_name='employees')
+
+# No
+from people.models import Person
+class Employee(models.Model):
+    person = models.ForeignKey(to=Person, on_delete=models.CASCADE)
+```
+
+### Related Name
+
+By default the related name of a relationship field is `{field_name}_set`,
+so by adding `related_name` parameter make the code cleaner and more readable.
+
+```python
+person = Person.objects.first()
+
+# Yes
+person.employees.all()
+
+# No
+person.employee_set.all()
+```
+
+## Model Controller
+
+Model Controller is a third party library that we use to keep track of data
+in each row, sometimes we would like to know when this row was create or
+update or who create or update this row.
+
+For installation and setup
+[read here](https://github.com/NorakGithub/django-model-controller) <sup>1</sup>
+
+```python
+from model_controller.models import AbstractSoftDeletionModelController
+
+
+class Person(AbstractSoftDeletionModelController):
+    pass
+```
+
+With this abstract class you get 5 fields:
+
+- `alive`: instead of delete row from database we just change the `alive` field to `null`, so that we can recover this row later.
+- `created_user/updated_user`: these fields are foreign key to `User` model. If you are using `ModelControllerSerializer` it will automatically add these field for you.
+- `created_at/updated_at`: these fields store the datetime fields when the record was created or latest update.
+
+Two important things you need to know after extended
+`AbstractSoftDeletionModelController`:
+
+- It overwrite the `delete()` method so that when you call delete method in your model object it will update `alive = null` instead of actually delete the object.
+- It provided you with two managers `objects` and `all_objects`
+
+```python
+# This will return person that alive is true
+Person.objects.all()
+
+# This will return all person rows in database include alive is null
+Person.all_objects.all()
+```
+
+### Unique
+
+When using `AbstractSoftDeletionModelController` if you want to make field value
+unique, make sure you use `unique_together` instead of field unique.
```

Comment:
should change to 

When using `AbstractSoftDeletionModelController` if you want to make field value
unique, make sure you use `unique_together` with `alive` field instead of field unique.

Reply:
[UniqueConstraint](https://docs.djangoproject.com/en/3.1/ref/models/constraints/#uniqueconstraint) should do the trick for this case krub.

Reply:
oh I just know that unique_together is deprcated. 😶

https://stackoverflow.com/questions/55495151/difference-between-uniqueconstraint-vs-unique-together-django-2-2

Reply:
Lol, I just know this one too